### PR TITLE
Tweak / fix deploy instructions

### PR DESCRIPTION
**Pull request by benthorner**
https://www.pivotaltracker.com/story/show/180900091

The original cherry-pick commit for manually testing changes no
longer works, so this updates it to a fresh version.

I also think it's worth includes the Hackney sample PDF for visual
checks as it's one of the few representative PDFs we have.

👉 [Rendered version](https://github.com/alphagov/notifications-template-preview/blob/tweak-deploy-instructions-180900091/docs/deploying.md)

👉 [Rendered version (new visual testing doc)](https://github.com/alphagov/notifications-template-preview/blob/tweak-deploy-instructions-180900091/docs/visual-testing.md)

### CODE REVIEW

**Review by leohemsted — COMMENTED** [review]
One comment, but a couple of other unrelated thoughts:

Do we want to encourage branch deploy, vs merging but being prepared to pin/roll back via concourse? The second doesn't involve us touching prod from our dev laptops, which I really don't like how much we do right now and am keen on avoiding. It also leaves an audit trail that should be clearer for other people to follow as they'll see the pinned resource in concourse. Some downsides are that it takes longer as you have to wait for preview and staging, and if you do need to revert then it clogs up the git history.

Is 4-5 a good time? While it might be quieter, it also means that if something goes wrong and we need to, for example, roll back and regenerate PDFs, or people saw errors and need to re-submit, there's not much time to do that. And in terms of business, today 4-5pm was very quiet but last monday it was actually the busiest hour of the day.

File: docs/deploying.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ Test the output looks visually OK before / after with a few sample PDFs. You sho
```
[the remark this answers is not in the document]
Comment (benthorner) [reply docs/deploying.md:20]: Thanks for spotting - I've amended the URL 👍.

This is a tricky one: these are really visual regression tests; if we want to keep them, then I'd like to make them proper automated tests and use a library to compare the before / after images automatically. However, I'm unsure if we want to keep them: doing these tests manually hasn't detected any issues since we introduced them, except in [the library to convert PDFs to PNGs](https://github.com/alphagov/notifications-template-preview/pull/591#issuecomment-979284671), which isn't critical. It felt reassuring to have them.

Historically all the issues we've had are with DVLA being unable to print a PDF due to a quirk of it's internals - I'm not aware of any visual hints of this, or any visual issues with PDFs in general. That said, when we're making a change like https://github.com/alphagov/notifications-template-preview/pull/632 and passing the entire PDF through a whole new mill, I do think we should be checking it hasn't done something awful to the output e.g. made it greyscale, deleted all the images, etc.

Short version:

- Do visual checks actually prevent scary issues at DVLA? No, I don't think so.
- Should we be visually checking PDFs when we make big changes? Probably.

Possible action:

- Move this guidance to a separate "visual testing" doc if we think it's discretionary.
- Suggest visual regression as an alternative approach on [this card](https://www.pivotaltracker.com/story/show/178746414).

What do you think?

**Comment by benthorner** [comment]
> Is 4-5 a good time?

Usually it's quieter, but not by much (looking at some recent stats). I'm up for removing it and agree the longer time to sort any immediate issues is an advantage over a slightly reduced average impact.

I've added a commit to relax this: https://github.com/alphagov/notifications-template-preview/pull/633/commits/dab335bc93a18be598dd33172b45e27c159b2815

> Do we want to encourage branch deploy, vs merging but being prepared to pin/roll back via concourse?

The complication here is the multi-day lag between doing a trial deployments and getting feedback from DVLA. If we merge as part of doing a trial deployment, then we'll be blocking deployments for quite a while.

I can't think of any good options here:

- Stop doing trial deployments. If https://github.com/alphagov/notifications-template-preview/pull/632 doesn't show up any issues, then I'm increasingly tempted by this one based. We have a recovery process, after all.

- Do high volume, representative load testing in Staging to emulate the trials in Production. This would involve analysing our Production PDFs in more detail to figure out what a good sample looks like.

Possible actions:

- Remove (or largely cut) this guidance if the redaction trial has no issues.

File: docs/deploying.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (leohemsted) [reply docs/deploying.md:20]: Yeah I definitely agree with you on those points. I don't think visual regression checks tie in with the thngs that DVLA struggle with but they still provide some value to us. We've historically had some issues where some PDFs we'd end up removing the logos when converting to CMYK IIRC? Others in the team might remember more. Not quite sure what it was - think we ended up bumping some libraries and that helped fix the issue. An easy way to test visually then might have helped.

Also, semi relatedly presumably flask shell is still a bit of a faff cos you need to open up a docker shell, run the stuff, work out where the files are, get them out with `docker cp`, compare them locally, etc. I think I ended up just writing a script using the python client to go via API at the time.

[the remark this answers is not in the document]
Comment (leohemsted) [reply docs/deploying.md:20]: as to your actions, i like moving it to a separate section. maybe reference that in this main bit and say "if you're worried your changes might impact pdf generation, consider running a manual visual regression test [ link to lower section ]"

**Comment by benthorner** [comment]
As promised, I've added a further commit to relax the guidance further: https://github.com/alphagov/notifications-template-preview/pull/633/commits/4038f9c0cd457d26147a1419815b1614f3b153c3

I think the question is now: do we want to make it entirely optional? I still think that would be unwise, given that we've no good way of doing E2E testing for this app, other than what the guidance says. If we did [this card](https://www.pivotaltracker.com/story/show/178746414), then I think we'd have a viable E2E testing strategy (albeit still a manual one) and removing these requirements would be safe.

**Review by leohemsted — APPROVED** [review]
Looks good, and I like the word suggested to imply we can use best wisdom but we probably should be thinking about it.

Good to go 👍

File: docs/deploying.md [hunk 2]
```diff
@@ -95,59 +61,23 @@ fields @timestamp, @message
...
-
-## Changes to previewing PDFs
-
-This isn't critical, so long as it works and doesn't differ much from the generated PDF. In the past we've found [there can be subtle before / after differences that aren't present in the PDF itself](https://github.com/alphagov/notifications-template-preview/pull/591#issuecomment-979284671).
-
-It's sufficient to test with one or two example PDFs e.g.
-
-- public_guardian_sample.pdf
-- example_dwp_pdf.pdf
+### Extended deployment
```
[the remark this answers is not in the document]
Comment (benthorner) [reply docs/deploying.md]: Fixed in https://github.com/alphagov/notifications-template-preview/pull/633/commits/0660ed2f742e22ac5059f93e546c62594fd63f15.
